File: day6/main.ts
import { readFileSync } from "fs";

function loadGrid(path: string): string[] {
  const lines = readFileSync(path, "utf8").split(/\r?\n/).filter((l) => l.length);
  const maxw = Math.max(...lines.map((l) => l.length));
  return lines.map((l) => l.padEnd(maxw, " "));
}

function splitBlocks(grid: string[]): Array<[number, number]> {
  const h = grid.length;
  const w = grid[0].length;
  const emptyCol: boolean[] = Array(w).fill(true);
  for (let c = 0; c < w; c++) {
    for (let r = 0; r < h; r++) {
      if (grid[r][c] !== " ") {
        emptyCol[c] = false;
        break;
      }
    }
  }
  const blocks: Array<[number, number]> = [];
  let c = 0;
  while (c < w) {
    while (c < w && emptyCol[c]) c++;
    if (c >= w) break;
    const start = c;
    while (c < w && !emptyCol[c]) c++;
    blocks.push([start, c]);
  }
  return blocks;
}

function problemOp(opRow: string): string {
  for (const ch of opRow) {
    if (ch === "+" || ch === "*") return ch;
  }
  return "+";
}

function evalNums(nums: bigint[], op: string): bigint {
  if (op === "+") return nums.reduce((a, b) => a + b, 0n);
  return nums.reduce((a, b) => a * b, 1n);
}

function part1(grid: string[], blocks: Array<[number, number]>): bigint {
  const opRow = grid[grid.length - 1];
  let total = 0n;
  for (const [s, e] of blocks) {
    const op = problemOp(opRow.slice(s, e));
    const nums: bigint[] = [];
    for (const row of grid.slice(0, -1)) {
      const tok = row.slice(s, e).trim();
      if (tok) nums.push(BigInt(tok));
    }
    total += evalNums(nums, op);
  }
  return total;
}

function part2(grid: string[], blocks: Array<[number, number]>): bigint {
  const h = grid.length - 1;
  const opRow = grid[grid.length - 1];
  let total = 0n;
  for (const [s, e] of blocks) {
    const op = problemOp(opRow.slice(s, e));
    const nums: bigint[] = [];
    for (let c = e - 1; c >= s; c--) {
      const digits: string[] = [];
      for (let r = 0; r < h; r++) {
        const ch = grid[r][c];
        if (/[0-9]/.test(ch)) digits.push(ch);
      }
      if (digits.length) nums.push(BigInt(digits.join("")));
    }
    total += evalNums(nums, op);
  }
  return total;
}

const grid = loadGrid("input.txt");
const blocks = splitBlocks(grid);
const t0 = performance.now();
const p1 = part1(grid, blocks);
const p2 = part2(grid, blocks);
const elapsed = performance.now() - t0;
console.log(
  `grand_total=${p1} quantum_total=${p2} elapsed_ms=${elapsed.toFixed(3)}`,
);

File: day6/main.rb
t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
grid = File.read('input.txt').split("\n").reject(&:empty?)
maxw = grid.map(&:length).max
grid.map! { |l| l.ljust(maxw, ' ') }

def split_blocks(grid)
  h = grid.size
  w = grid.first.size
  empty = Array.new(w, true)
  (0...w).each do |c|
    (0...h).each do |r|
      if grid[r][c] != ' '
        empty[c] = false
        break
      end
    end
  end
  blocks = []
  c = 0
  while c < w
    c += 1 while c < w && empty[c]
    break if c >= w
    start = c
    c += 1 while c < w && !empty[c]
    blocks << [start, c]
  end
  blocks
end

def problem_op(s)
  s.each_char do |ch|
    return ch if ch == '+' || ch == '*'
  end
  '+'
end

def eval_nums(nums, op)
  if op == '+'
    nums.sum
  else
    nums.reduce(1) { |a, b| a * b }
  end
end

def part1(grid, blocks)
  op_row = grid[-1]
  total = 0
  blocks.each do |s, e|
    op = problem_op(op_row[s...e])
    nums = []
    grid[0...-1].each do |row|
      tok = row[s...e].strip
      nums << tok.to_i unless tok.empty?
    end
    total += eval_nums(nums, op)
  end
  total
end

def part2(grid, blocks)
  h = grid.size - 1
  op_row = grid[-1]
  total = 0
  blocks.each do |s, e|
    op = problem_op(op_row[s...e])
    nums = []
    (e - 1).downto(s) do |c|
      digits = []
      (0...h).each do |r|
        ch = grid[r][c]
        digits << ch if ch =~ /[0-9]/
      end
      nums << digits.join.to_i unless digits.empty?
    end
    total += eval_nums(nums, op)
  end
  total
end

blocks = split_blocks(grid)
p1 = part1(grid, blocks)
p2 = part2(grid, blocks)
elapsed_ms = (Process.clock_gettime(Process::CLOCK_MONOTONIC) - t0) * 1000.0
puts "grand_total=#{p1} quantum_total=#{p2} elapsed_ms=#{format('%.3f', elapsed_ms)}"
